Add jenkinsfile

richardbadman opened the pull request:
This adds jenkins support for the solr repo, and will do the following:

1. Push an ant-based docker image to our gcp repo _(under `ant-base`)_
2. Push a solr based docker image to our gcp repo _(under `bw-lucene-solr`)_

This was mainly done to make the solr prometheus exporter dockerising significantly easier.

The ant-base image doesn't need to change much, only for each major version of solr, because of this, there is a check to see if there is an image for that major version of solr in our gcp already, if there is, it'll use that, otherwise create one.

Version numbering is maintained in a `versions.json` file, I originally had some regex stuff to not need a `versions.json` file, but then it could get complicated with the relying of branch naming convention.

richardbadman committed:
Add base Jenkinsfile to repo

richardbadman committed:
Fleshed out Jenkinsfile

richardbadman committed:
Add newer refreshed dockerfile

richardbadman committed:
Add build phase

This should hopefully build the docker image

richardbadman committed:
Made a base docker file for ant

This makes an ant image which can be used for building the solr packages
quicker.

Work is still to be done to append it to latest major version of solr
and implement checks for it

Fix main dockerfile to work and use ant image

This now uses the ant image to make it a lot easier and quicker to build

richardbadman committed:
Make images push to gcr

This should hopefully push to our private repo now

NT added comments

richardbadman committed:
Use versions.json instead

Instead of the funky regex, and relying on the branch name, just going
ot use a versions.json instead

richardbadman committed:
Made function to push images

Cleaned up the code a bit to have a function to push images instead

richardbadman committed:
Fix for functions

jward-bw commented on Dockerfile:
```diff
@@ -1,76 +1,8 @@
-FROM debian:jessie
+ARG REPO=""
```
Is having `""` default necessary?

richardbadman replied:
Tbf I didn't try just having `ARG REPO`

jward-bw commented on Jenkinsfile:
```diff
@@ -0,0 +1,76 @@
+#!groovy
+gcpProject =  "bw-prod-platform0"
+baseImageExist = false
```
Is this default necessary?

jward-bw commented on Jenkinsfile:
```diff
@@ -0,0 +1,76 @@
+#!groovy
+gcpProject =  "bw-prod-platform0"
```
NABD - extra space after `=`

jward-bw commented on Jenkinsfile:
```diff
@@ -0,0 +1,76 @@
+                    --build-arg MAJOR_VERSION=${major_version} \
+                    -t ${gcpProject}/bw-lucene-solr:${version} .
+                """
+                pushImage("${gcpProject}/bw-lucene-solr:${version}", version)
```
I would recommend only pushing if you're on the master branch. Otherwise you can overwrite an image just by raising a PR. See: https://github.com/BrandwatchLtd/bw-airflow/blob/db6c449f050d4a507005f7feb96af74e1028fc0f/Jenkinsfile#L47

richardbadman replied:
The branches don't work like that for solr, you have;
```
[apache branches]
branch_x_y(_z)
i.e.
branch_7_7_2
branch_8_4

[bw branches (patches/custom stuff applied)]
bw_branch_x_y(_z)
ie.e.
bw_branch_7_7_2
bw_branch_8_4
```

Master on bw side, is massively out of date, and on apache is the most recent, often un-released versions of solr

jward-bw replied:
In that case you can check if you're on a pull request with:
```groovy
if (!env.CHANGE_ID) {
    // not a PR, so can push the images
}

jward-bw replied:
https://jenkins.brandwatch.net/job/BrandwatchLtd/job/bw-lucene-solr/pipeline-syntax/globals#env

richardbadman replied:
I just thought, I could push another tag which is the build number, that would probs be safest

jward-bw replied:
Yeah I think a lot of bw repos use `build.${build-number}` or something to that effect

jward-bw commented on Jenkinsfile:
```diff
@@ -0,0 +1,76 @@
+                sh "docker build -f base.Dockerfile -t ${gcpProject}/ant-base:${major_version} ."
+                pushImage("${gcpProject}/ant-base:${major_version}", major_version)
+
+                sh """ docker build \
+                    -f Dockerfile \
+                    --build-arg REPO=${gcpProject} \
+                    --build-arg MAJOR_VERSION=${major_version} \
+                    -t ${gcpProject}/bw-lucene-solr:${version} .
+                """
+                pushImage("${gcpProject}/bw-lucene-solr:${version}", version)
```
Code duplication, could change this around to:
```groovy
if ( !baseImageExists ) {
    // build ant image
}
// build bw-lucene-solr image

jward-bw commented on Jenkinsfile:
```diff
@@ -0,0 +1,76 @@
+            echo "Base image exists, using that"
+            withEnv(["DOCKER_BUILDKIT=1"]) {
+                sh """ docker build \
+                    -f Dockerfile \
```
NABD - `-f Dockerfile` is unnecessary as docker looks for a file `Dockerfile` by default.

jward-bw commented on base.Dockerfile:
```diff
@@ -0,0 +1,44 @@
+FROM debian:stretch
```
Is there any reason you can't just use the prebuilt openjdk base image? E.g. `openjdk:8-stretch`:
https://hub.docker.com/layers/openjdk/library/openjdk/8-stretch/images/sha256-6896e96ce729fd54da25b8d831400773efeef39780c65bffb2d64c1412189455?context=explore

richardbadman replied:
I'll try that image and see if it works

jward-bw commented on base.Dockerfile:
```diff
@@ -0,0 +1,44 @@
+RUN fpm --version
+
+#Installing Apache Ant
+RUN wget http://archive.apache.org/dist/ant/binaries/apache-ant-1.10.5-bin.tar.gz
```
Do you need this particular version of ant? You could consider having the ant version in your `versions.json` and tagging the ant image with that version number.

richardbadman replied:
This was the version of ant I installed originally, I haven't really tested any other versions

jward-bw replied:
The main reason I said this is it feels weird that you're rebuilding the ant image based on the solr major version, when it has no dependencies on that. Realistically you probably want to be rebuilding this image every time you merge to pick up security fixes.

richardbadman replied:
🤷 I'm not entirely sure, I can't remember how I installed ant on my machine _(unless it was pre installed)_, and back in the day before jenkins was updated, we had a node which just had things hard installed on there , so I had to spend time just trying to figure out how to install ant on a docker image to build jenkins

jward-bw replied:
Every merge may be a bit too keen actually if this repo is updated frequently.

richardbadman replied:
Following your other comment, I'm re-working this to make a simpler ant image, and having a parameter for versions, not sure which version is best to use, but will stick with the current version we've been running for now _(but ofc, has the option to change in future)_

jward-bw commented on Jenkinsfile:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    stage ('Build images') {
+        baseImageExist = doesBaseImageExist(major_version)
```
NABD - `baseImageExists`

richardbadman committed:
Made suggested changes

Made the suggested changes, most stylistic, but have also done the
following:

  - Changed the base image to be simpler, and easily maintainable
  - Added support for ant versioning
  - Added build number to images
  - Changed solr build process args which should be quicker

jward-bw reviewed (commented):
What are you using the bw-lucene-solr image for? If you are using it for any production purposes, or building any production images based on that image, then you need to be building the base ant image regularly to fix any security vulnerabilities.

richardbadman committed:
Force ant to build every time

richardbadman commented:
@jward-bw I've made it build ant every time now, however, the security issues you pointed out are still persistent with the current version of openjdk and this specific ant version.

Like I said previously though, I don't wish to spend any time further investigating this to find a secure ant version unless it's something quick and easy

jward-bw commented:
> @jward-bw I've made it build ant every time now, however, the security issues you pointed out are still persistent with the current version of openjdk and this specific ant version.
> 
> Like I said previously though, I don't wish to spend any time further investigating this to find a secure ant version unless it's something quick and easy

No that's good, all this will do is make sure the next time you build the image it picks up the security fixes (when someone implements them upstream). We could do something more like running this Jenkinsfile regularly to update the images, but I think this is fine for now.
